pkg/controller/attach_detach.go

```diff
@@ -573,154 +559,162 @@ func reportError(azVolumeAttachment *azdiskv1beta2.AzVolumeAttachment, state azd
+	if err != nil {
+		w.Logger().Errorf(err, "failed to get list of nodes")
+		return err
+	}
 
-				for k, v := range desiredAzVolumeAttachment.Labels {
-					azVolumeAttachment.Labels = azureutils.AddToMap(azVolumeAttachment.Labels, k, v)
-				}
+	for _, node := range nodes.Items {
+		nodeName := node.Name
+		// Get all attached disks from VM
+		disks, _, err := r.cloudDiskAttacher.GetNodeDataDisks(types.NodeName(nodeName))
+		if err != nil {
+			w.Logger().Errorf(err, "failed to get data disks for node (%s).", nodeName)
+			continue
+		}
 
-				for k, v := range azVolumeAttachment.Annotations {
-					azVolumeAttachment.Status.Annotations = azureutils.AddToMap(azVolumeAttachment.Status.Annotations, k, v)
-				}
+		for _, dataDisk := range disks {
+			volumeName := strings.ToLower(*dataDisk.Name)
+			azVolumeAttachmentName := azureutils.GetAzVolumeAttachmentName(volumeName, nodeName)
+			err = r.recreateAzVolumeAttachmentFromVolumeInfo(ctx, "", azVolumeAttachmentName, volumeName, nodeName)
+			w.Logger().Errorf(err, "failed to recreate replica AzVolumeAttachment (%s) from volumeInfo.", azVolumeAttachmentName)
+		}
+	}
+	return err
+}
 
-				// for now, we don't empty the meta annotatinos after migrating them to status annotation for safety.
-				// note that this will leave some remnant garbage entries in meta annotations
+func generateVolumeInfo(volumeHandle string, volumeContext map[string]string) volumeInfo {
+	var volumeInfo volumeInfo
+	volumeInfo.handle = volumeHandle
+	volumeInfo.attributes = volumeContext
+	volumeInfo.syncedAzVolumeAttachments = make(map[string]bool)
+	return volumeInfo
+}
 
-				for k, v := range desiredAzVolumeAttachment.Annotations {
-					azVolumeAttachment.Annotations = azureutils.AddToMap(azVolumeAttachment.Status.Annotations, k, v)
-				}
+func (r *ReconcileAttachDetach) loadOrStoreVolumeInfo(ctx context.Context, volumeName, azVolumeAttachmentName string) (string, map[string]string, volumeInfo, bool, error) {
+	w, _ := workflow.GetWorkflowFromContext(ctx)
+	var err error
+	var volumeHandle string
+	var volumeContext map[string]string
+	var vInfo = volumeInfo{}
+	hasSynced := false
+	if val, ok := r.volumeInfos.Load(volumeName); ok {
+		vInfo = val.(volumeInfo)
+		if synced, hasAzVolumeAttachment := vInfo.syncedAzVolumeAttachments[azVolumeAttachmentName]; hasAzVolumeAttachment && synced {
+			hasSynced = true
+			return volumeHandle, volumeContext, vInfo, hasSynced, nil
+		}
+		volumeHandle = vInfo.handle
+		volumeContext = vInfo.attributes
+	} else {
+		w.Logger().Infof("getting volumeHandle and volumeContext from PersistentVolume for volume (%s)", volumeName)
+		volumeHandle, volumeContext, err = r.getVolumeInfoFromPV(ctx, volumeName)
+		if err != nil {
+			w.Logger().Errorf(err, "failed to find volume info from PersistentVolume for volume (%s)", volumeName)
+			return "", nil, vInfo, hasSynced, err
+		}
+		vInfo = generateVolumeInfo(volumeHandle, volumeContext)
+		r.volumeInfos.Store(volumeName, vInfo)
```

Comment:
In general when using `sync.Map`, you need to store using `LoadOrStore`. Otherwise, you risk overwriting the data another goroutine has added.

The general usage pattern for `sync.Map` is:

```go
entry, ok := syncMap.Load(key)
if !ok {
    entry, _ := syncMap.LoadOfStore(key, newEntry())
}
```